modules/runtime/nuxeo-runtime-metrics/src/main/java/org/nuxeo/runtime/metrics/reporter/JaegerReporter.java

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * @since 11.1
+ */
+public class JaegerReporter extends AbstractMetricsReporter {
+
+    private static final Logger log = LogManager.getLogger(JaegerReporter.class);
+
+    public static final String URL = "url";
+
+    public static final String TIMEOUT_SECOND = "timeoutSecond";
+
+    public static final String DEFAULT_TIMEOUT_SECOND = "10";
+
+    public static final String MAX_ATTRIBUTES = "maxAttributes";
+
+    public static final String DEFAULT_MAX_ATTRIBUTES = "128";
+
+    public static final String MAX_ANNOTATIONS = "maxAnnotations";
+
+    public static final String DEFAULT_MAX_ANNOTATIONS = "128";
+
+    public static final String SAMPLER_PROB = "samplerProbability";
+
+    public static final String DEFAULT_SAMPLER_PROB = "0.1";
+
+    protected boolean activated;
+
+    @Override
+    public void start(MetricRegistry registry, MetricFilter filter, Set<MetricAttribute> deniedExpansions) {
+        log.warn("Creating Jaeger reporter");
+        String url = options.get(URL);
+        Duration timeout = Duration.create(Long.valueOf(options.getOrDefault(TIMEOUT_SECOND, DEFAULT_TIMEOUT_SECOND)),
+                0);
+        JaegerExporterConfiguration configuration = JaegerExporterConfiguration.builder()
+                                                                               .setServiceName("nuxeo")
+                                                                               .setThriftEndpoint(url)
+                                                                               .setDeadline(timeout)
+                                                                               .build();
+        JaegerTraceExporter.createAndRegister(configuration);
+        activated = true;
+        enableTracing(options);
+    }
+
+    public static void enableTracing(Map<String, String> options) {
+        TraceConfig traceConfig = Tracing.getTraceConfig();
+        TraceParams activeTraceParams = traceConfig.getActiveTraceParams();
+        TraceParams.Builder builder = activeTraceParams.toBuilder();
+        int maxAttributes = Integer.parseInt(options.getOrDefault(MAX_ATTRIBUTES, DEFAULT_MAX_ATTRIBUTES));
+        int maxAnnotations = Integer.parseInt(options.getOrDefault(MAX_ANNOTATIONS, DEFAULT_MAX_ANNOTATIONS));
+        builder.setMaxNumberOfAttributes(maxAttributes).setMaxNumberOfAnnotations(maxAnnotations);
+        float samplerProbability = Float.valueOf(options.getOrDefault(SAMPLER_PROB, DEFAULT_SAMPLER_PROB));
+        if (samplerProbability >= 0.999) {
+            builder.setSampler(Samplers.alwaysSample());
+        } else if (samplerProbability <= 0.001) {
+            builder.setSampler(Samplers.neverSample());
+        } else {
+            builder.setSampler(Samplers.probabilitySampler(samplerProbability));
+        }
+        traceConfig.updateActiveTraceParams(builder.build());
+    }
+
+    @Override
+    public void stop() {
+        log.debug("Stop reporting");
+        if (activated) {
```

Comment:
Agree, the runtime guarantees the activate/start/stop/deactivate lifecycle and each stage is called only once per component.

Here, the only benefit I saw to have the activated boolean is to make these classes stateless regarding the error case. Indeed if start stage fails but don't crash the runtime, the stop stage will be called. We may want to have this try/catch mechanism on the component starting these reporters and not register them if so, as it during stop we will stop only the active ones?

Reply:
I have no way to check if the exporter is started so I use a boolean to make sure unregister is called only when exporter start succeed. In case of failure during start, it prevents to get also a failure during stop.